core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -59,40 +78,115 @@ object ByteBufferMessageSet {
     }
   }
 
-  /** Deep iterator that decompresses the message sets in-place. */
-  def deepIterator(wrapperMessage: Message): Iterator[MessageAndOffset] = {
+  /** Deep iterator that decompresses the message sets and adjust timestamp and offset if needed. */
+  def deepIterator(wrapperMessageAndOffset: MessageAndOffset): Iterator[MessageAndOffset] = {
+
+    import Message._
+
     new IteratorTemplate[MessageAndOffset] {
 
+      val wrapperMessageOffset = wrapperMessageAndOffset.offset
+      val wrapperMessage = wrapperMessageAndOffset.message
+      val wrapperMessageTimestampOpt: Option[Long] =
+        if (wrapperMessage.magic > MagicValue_V0) Some(wrapperMessage.timestamp) else None
+      val wrapperMessageTimestampTypeOpt: Option[TimestampType] =
+        if (wrapperMessage.magic > MagicValue_V0) Some(wrapperMessage.timestampType) else None
+      if (wrapperMessage.payload == null)
+        throw new RuntimeException("wrapper message = " + wrapperMessage)
       val inputStream: InputStream = new ByteBufferBackedInputStream(wrapperMessage.payload)
       val compressed: DataInputStream = new DataInputStream(CompressionFactory(wrapperMessage.compressionCodec, inputStream))
-
-      override def makeNext(): MessageAndOffset = {
+      var lastInnerOffset = -1L
+      // When magic value is greater than 0, relative offset will be used.
+      // Ideally the conversion from relative offset(RO) to absolute offset(AO) should be:
+      //
+      // AO = AO_Of_Last_Inner_Message + RO
+      //
+      // However, Note that the message sets sent by producers are compressed in a stream compressing way.
+      // And the relative offset of an inner message compared with the last inner message is not known until
+      // the last inner message is written.
+      // Unfortunately we are not able to change the previously written messages after the last message is writtern to
+      // the message set when stream compressing is used.
+      //
+      // To solve this issue, we use the following solution:
+      // 1. When producer create a message set, it simply write all the messages into a compressed message set with
+      //    offset 0, 1, ... (inner offset).
+      // 2. The broker will set the offset of the wrapper message to the absolute offset of the last message in the
+      //    message set.
+      // 3. When a consumer sees the message set, it first decompresses the entire message set to find out the inner
+      //    offset (IO) of the last inner message. Then it computes RO and AO of previous messages:
+      //
+      //    RO = Inner_Offset_of_a_message - Inner_Offset_of_the_last_message
+      //    AO = AO_Of_Last_Inner_Message + RO
+      //
```

Comment:
Could we put those comments in a more prominent place like the beginning of the class? With v1 message format, we are adding a timestamp, a timestamp type attribute, and are using a relative for inner message. It would be useful to document the format in a bit more details for both the outer and the inner message. For example, should the timestamp type attribute be set for inner messages?